src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnectionPool.cs

```diff
@@ -1135,34 +1041,47 @@ public CachedConnection(HttpConnection connection)
         /// into the waiter as a result, and if no connection is available from the pool,
         /// this waiter's logic is used to create the connection.
         /// </summary>
-        private class ConnectionWaiter : TaskCompletionSource<(HttpConnectionBase connection, bool isNewConnection, HttpResponseMessage failureResponse)>
+        private sealed class ConnectionWaiter
         {
-            /// <summary>The pool with which this waiter is associated.</summary>
-            internal readonly HttpConnectionPool _pool;
-            /// <summary>Request to use to create the connection.</summary>
-            private readonly HttpRequestMessage _request;
-
-            /// <summary>Cancellation token for the waiter.</summary>
-            internal readonly CancellationToken _cancellationToken;
-            /// <summary>Registration that removes the waiter from the registration list.</summary>
-            internal CancellationTokenRegistration _cancellationTokenRegistration;
+            private readonly TaskCompletionSource<HttpConnection> _tcs;
+
             /// <summary>Next waiter in the list.</summary>
             internal ConnectionWaiter _next;
             /// <summary>Previous waiter in the list.</summary>
             internal ConnectionWaiter _prev;
 
             /// <summary>Initializes the waiter.</summary>
-            public ConnectionWaiter(HttpConnectionPool pool, HttpRequestMessage request, CancellationToken cancellationToken) : base(TaskCreationOptions.RunContinuationsAsynchronously)
+            public ConnectionWaiter()
             {
-                Debug.Assert(pool != null, "Expected non-null pool");
-                _pool = pool;
-                _request = request;
-                _cancellationToken = cancellationToken;
+                _tcs = new TaskCompletionSource<HttpConnection>(TaskCreationOptions.RunContinuationsAsynchronously);
```

Comment:
This approach is a bit more expensive than was here previously:
- Now we have two allocations (the ConnectionWaiter and the TCS) instead of one (the ConnectionWaiter that is a TCS).
- The GetConnectionAsync method below will end up allocating an async state machine / Task in addition to this TCS' Task, whereas before the await happened on the TCS Task.

Reply:
That's true, but I don't know if it really matters in practice.

I could change the ConnectionWaiter to inherit from TCS again; the main reason not to is just to ensure that we don't accidentally access the TCS directly.

Do you think this is worth changing?

Reply:
As I mentioned above, one possibility to reduce memory allocations here is:

> We could go further here and just use a Queue<TaskCompletionSource<HttpConnection>> instead of the doubly linked list we have today.

Reply:
We had the doubly-linked list in order to support efficient removal from the middle of the list, in support of cancellation, and because we already needed a TCS, we could just create ConnectionWaiter to derive from it and add the links, allowing us to reuse the same object we already needed as the node in the list, for limited additional overhead.

Now that we no longer need removal from the middle of the list, switching to a `Queue<TaskCompletionSource>` seems good.